newIDE/app/src/CommandPalette/CommandManager.js

```diff
@@ -3,21 +3,49 @@ import { type MessageDescriptor } from '../Utils/i18n/MessageDescriptor.flow';
 
 type CommandHandler = () => void | Promise<void>;
 
-export type Command = {|
+export type SimpleCommand = {|
   displayText: MessageDescriptor,
   enabled: boolean,
   handler: CommandHandler,
 |};
 
+export type CommandOption<T> = {|
+  value: T,
+  handler: CommandHandler,
+  text: string,
+  iconSrc?: string,
+|};
+
+export type CommandWithOptions<T> = {|
+  displayText: MessageDescriptor,
+  enabled: boolean,
+  generateOptions: () => Array<CommandOption<T>>,
+|};
+
+export type Command = SimpleCommand | CommandWithOptions<*>;
+
 export type NamedCommand = {|
   name: string,
   ...Command,
 |};
 
-export default class CommandManager {
+export type NamedCommandWithOptions<T> = {|
+  name: string,
+  ...CommandWithOptions<T>,
+|};
+
+export interface CommandManagerInterface {
+  commands: { [string]: Command };
```

Comment:
Are you forced to include commands in the interface? Typically we don't want to expose this to the client code.